File: src/main/java/henaxel/raytracer/Sphere.java
package henaxel.raytracer;

import henaxel.raytracer.utils.Vec3;
import henaxel.materials.Material;

public class Sphere extends Surface {
    double radius;

    public Sphere(Vec3 pos, double radius, Material material) {
        super(pos, material);
        this.radius = radius;
    }

    @Override
    protected HitResult hit(Ray r, double t_min, double t_max) {
        Vec3 oc = Vec3.subtract(r.origin(), pos);
        double a = Vec3.dot(r.direction(), r.direction());
        double b = Vec3.dot(oc, r.direction());
        double c = Vec3.dot(oc, oc) - radius*radius;
        double discriminant = b*b - a*c;

        if(discriminant > 0) {
            double t = (-b - Math.sqrt(discriminant))/a;
            if (t < t_max && t > t_min) {
                Vec3 hitPos = r.pointAtParameter(t);
                Vec3 n = Vec3.divide(Vec3.subtract(hitPos, pos), radius);
                Ray scatter = mat.scatter(r, hitPos, n);
                
                double[] uv = getUV(hitPos);
                return new HitResult(hitPos, n, t, scatter, mat.getAlbedo(uv[0], uv[1])); //TODO : Implement proper uv mapping
            }
            t = (-b + Math.sqrt(discriminant))/a;
            if (t < t_max && t > t_min) {
                Vec3 hitPos = r.pointAtParameter(t);
                Vec3 n = Vec3.divide(Vec3.subtract(hitPos, pos), radius);
                Ray scatter = mat.scatter(r, hitPos, n);

                double[] uv = getUV(hitPos);
                return new HitResult(hitPos, n, t, scatter, mat.getAlbedo(uv[0], uv[1])); //TODO : Implementing proper uv mapping
            }

        }
        return null;
    }
    
    @Override
    protected double[] getUV(Vec3 p) {
        p = Vec3.divide(Vec3.subtract(p, pos), radius);
        double[] uv = new double[2];
        double phi = Math.atan2(p.z, p.x);
        double theta = -Math.asin(p.y);
        uv[0] = 1-(phi + Math.PI) / (2*Math.PI);
        uv[1] = (theta + Math.PI/2) / Math.PI;
        return uv;
    }
    
    @Override
    protected AABB boundingBox() {
        return new AABB(Vec3.subtract(pos, new Vec3(radius, radius, radius)), Vec3.add(pos, new Vec3(radius, radius, radius)));
    }
}

File: src/henaxel/raytracer/Sphere.kt
package henaxel.raytracer

import henaxel.raytracer.materials.Material
import henaxel.raytracer.utils.Vec3
import kotlin.math.PI
import kotlin.math.asin
import kotlin.math.atan2
import kotlin.math.sqrt

open class Sphere(pos: Vec3, var radius: Double, material: Material) : Surface(pos, material) {
    override fun hit(ray: Ray, tMin: Double, tMax: Double): HitResult? {
        val oc = ray.pos - pos
        val a = ray.dir * ray.dir
        val b = oc * ray.dir
        val c = (oc * oc) - radius*radius
        val discriminant = b*b - a*c

        if (discriminant > 0) {
            var t = (-b - sqrt(discriminant)) / a
            if (!(t < tMax && t > tMin)) {
                t = (-b + sqrt(discriminant)) / a
                if (!(t < tMax && t > tMin)) return null
            }
            val hitPos = ray.pointAtParameter(t)
            val normal = (hitPos-pos)/radius
            val scatter = mat.scatter(ray, hitPos, normal)
            val uv = getUV(hitPos)

            return HitResult(hitPos, normal, t, scatter, mat.getAlbedo(uv[0], uv[1]))
        }
        return null
    }

    override fun getUV(p: Vec3): Array<Double> {
        val p = (p - pos)/radius
        val phi = atan2(p.z, p.x)
        val theta = -asin(p.y)
        return arrayOf(
            1-(phi + PI) / (2*PI),
            (theta + PI/2) / PI
        )
    }

    override fun boundingBox(): AABB {
        return AABB(pos - Vec3(radius,radius,radius),pos + Vec3(radius,radius,radius))
    }
}